### PR TITLE
Revamped the find&replace UI.

### LINKED ISSUE #9928: Stylelint verification issues
Currently, there are some issues in the `findandreplaceform.css` file being highlighted by stylelint, for example:

```
51:2 ✖ Expected selector ".ck .ck-find-and-replace-form .ck-replace-form__wrapper .ck-button"
to come before selector ".ck .ck-find-and-replace-form .ck-find-form__wrapper .ck-button:hover"
no-descending-specificity
```

This needs to be reviewed and cleaned up.

### LINKED ISSUE #10019: The find and replace view looks un-styled when there is not enough space (responsive design)
## 📝 Provide detailed reproduction steps (if any)

1.  http://fake.ckeditor.com:8081/build/docs/ckeditor5/latest/features/find-and-replace.html
2.  shrink the window width
3.  open the find and replace view

### ✔️ Expected result

UI should stay readable

### ❓ Possible solution

_If you have ideas, you can list them here. Otherwise, you can delete this section._

## 📃 Other details

*   Browser: …
*   OS: …
*   First affected CKEditor version: …
*   Installed CKEditor plugins: …

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### LINKED ISSUE #10021: The find and replace unexpected behavior on toggling "match case"
## 📝 Provide a description of the improvement

If I toggle the "match case" button the state of other buttons change but markers are still there, and when I toggle it back the UI switches to the previous state. The same for the input field - when changed, only the UI state changes but not occurrences, and while changed back to previous input, the UI gets back to the previous state. 

I'm not sure what should be the best UX but I guess the search results should clear every time the input field or checkbox state changes and should not be brought back after getting back to the previous state.

## 📃 Other details

*   Browser: …
*   OS: …
*   CKEditor version: …
*   Installed CKEditor plugins: …

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### LINKED ISSUE #10229: Visually refine find and replace panel
### Current design

It stands out and does not follow the look&feel of other features.

Note: This issue includes both design and implementation.

### Scope

*   Design a new look&feel of the F&R panel based on the feedback that we got.
*   Implement the new designs into the existing F&R code.
    *   Next/prev buttons will be permanently visible
    *   Move "match case", "whole..." options to a dropdown
    *   Make it wider
    *   Style all buttons (use existing classes)
    *   We ignore smaller screen sizes for now (no responsiveness to be included in this ticket) because the feature is useless on mobiles anyway as it takes the entire screen size
    *   Consider different translations (length of text in the buttons like "Find")
    *   Add header "Find and replace"
        *   TODO: Is it a general component or does it need to be moved from the Special characters balloon
        *   Answer: We moved it to ckeditor5-ui already so it can be reused
    *   ~Replace the split button used in the toolbar with a normal button~ Done already
    *   Missing find text error (tooltip) – use an existing input property
    *   New state logic for buttons (which are enabled/disabled when)

### Note

*   F&R is disabled on CI as it has incomplete CC (see #10290)

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### LINKED ISSUE #10304: Changing the search text should reset the results
## 📝 Provide detailed reproduction steps (if any)

1.  Type "Steam"
2.  Click "Find"
3.  Press backspace

### ✔️ Expected result

Highlights are removed. The state of the form gets back to its initial state.

### ❌ Actual result

The results for "Steam" are still displayed.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (find-and-replace): Visually revamped the find&replace form. Closes #10229.

Fix (find-and-replace): Changing the search text should reset the results. Closes #10304.

Fix (find-and-replace): Toggling search options should reset the results. Closes #10021.

Fix (find-and-replace): The find and replace form should be responsive. Closes #10019.

Internal (find-and-replace): Fixed the styling issue that was triggering style linter. Removed linter's ignore clause. Closes #9928.

Feature (core): Added the "cog" icon to the core icon set (see #10229).

Other (theme-lark): Moved presentational find and replace form styles to the theme (see #10229).

Fix (theme-lark): The label of the labeled field should stay on the top when the field is disabled and not empty to not cover the field's text (see #10229).

BREAKING CHANGE (find-and-replace): The layout, styling, and view structure of the find and replace form have changed radically, which may affect integrations that either customized or extended this form (see #10229). 

### Additional information

The list of things in this PR:

*   Complete revamp of the UI according to the designs in the issue,
*   100% CC for `FindAndReplace`, `FindAndReplaceUI`, `FindAndReplaceFormView`,
*   Code refactoring in mentioned classes to decouple data flow/application layers that got tangled in some places.
*   General code refactoring where necessary.
*   Complete API documentation in mentioned classes.
*   Decoupled wireframe and presentational styles of the form.
*   RWD support.
*   RTL support.
*   Smart counter management to avoid situations like below where the input text touches the counter or gets covered by the counter (the text in the screenshots is "A steam locomotive is one that uses a steam engine as the prime source of power to move and pull the cars.").

or

or

### CODE REVIEW

**Comment by oleq** [comment]
@ckeditor/qa-team Can you take a look, please? 🙏 We're looking for regressions and glitches in the form. The engine part of the feature was untouched.

**Comment by LukaszGudel** [comment]
### Steps:
1. Open F&R dropdown and remove focus from "find text" input.
2. Drag and drop some text from another window into the input. 

### Result:
Placeholder/label does not change its position and now text in the input overlaps with the label:

https://user-images.githubusercontent.com/58658048/129527292-5e92bf8d-5d89-40a6-ae53-ce9f00a33359.mp4

### Notes:
There are differences depending on the browser:
- Chrome - result as shown on the video, focusing the input is fixing the issue
- Firefox - result as shown on the video, but focusing the input and removing the focus will still show overlapped texts. You can clear the input and type something to fix it
- Safari - issue does not occur, dropping the text into the input is correctly changing the position of the label

**Comment by LukaszGudel** [comment]
### Steps [SAFARI only]:
2. Open F&R dropdown.
3. Click on it.

### Result:
There is a blue outline with an irregular shape:


https://user-images.githubusercontent.com/58658048/129528813-1d1f5a9f-16fc-4c59-838d-8d72979986e3.mp4

**Comment by Reinmar** [comment]
The currently selected result is red. It was orange on `master`. Is this an intentional change? We have #10242 but this PR doesn't state that it's going to be closed. It wasn't part of this ticket either.

image

**Comment by Reinmar** [comment]
This PR resolves #10304. To me, it also resolves #10021.

**Comment by Reinmar** [comment]
> The currently selected result is red. It was orange on `master`. Is this an intentional change? We have #10242 but this PR doesn't state that it's going to be closed. It wasn't part of this ticket either.

Let's agree to revert that change as none of us think it was intentional and then fix this in #10242.

**Comment by pomek** [comment]
Do not forget about executing tests locally as we don't execute these for the find-and-replace feature on CI (https://github.com/ckeditor/ckeditor5/issues/9996).

```
FAILED TESTS:
  FindAndReplaceFormView
    form state machine
      using the replace UI
        ✖ should display a tip when the replace field is disabled and focused
          Chrome 92.0.4515.131 (Mac OS 10.15.7)
        AssertionError: expected '' to match /^Tip: Find some text/

      + expected - actual


    at Context.eval (webpack:///./packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js?:1097:40)
```

**Comment by Reinmar** [comment]
@maxbarnas, since Olek is off for this week feel free to introduce changes to this PR while reviewing it.

**Comment by Reinmar** [comment]
@LukaszGudel, thanks for the testing.

The [first issue that you found](https://github.com/ckeditor/ckeditor5/pull/10344#issuecomment-899293391) is not related to F&R itself. It's probably something with our inputs.

The second one is not critical as well.

Both deserve to be extracted to new issues, but I consider both low priority.

**Comment by Reinmar** [comment]
@maxbarnas Could I also ask you to verify issues from this epic: #10290? Which of them can be closed with this PR and which needs to be addressed later?

**Comment by Reinmar** [comment]
image

**Comment by Reinmar** [comment]
Will also resolve #10019.

**Comment by maxbarnas** [comment]
First thing that caught my eye is the fact that the F&R popup can cover the search results. Best example:
image
Can you tell me where is the 7th "chocolate" here? ;) Closing the F&R popup also removes the highlight :(

File: packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js [hunk 1]
```diff
@@ -67,333 +355,889 @@ describe( 'FindAndReplaceFormView', () => {
...
+			} );
+
+			it( 'should display a tip when the replace field is disabled but not focused', () => {
+				openDropdown();
+
+				expect( replaceInput.isEnabled ).to.be.false;
+				expect( replaceInput.infoText ).to.equal( '' );
+			} );
+
+			it( 'should display a tip when the replace field is disabled and focused', () => {
```
Comment (maxbarnas) [inline packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js:1090]: This test fails when the Chrome window, that executes the tests, is not focused. @oleq, AFAIK, you created a workaround some time ago for a similar problem, right?

File: packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js [hunk 2]
```diff
@@ -67,333 +355,889 @@ describe( 'FindAndReplaceFormView', () => {
...
+				openDropdown();
+
+				replaceInput.focus();
+
+				expect( replaceInput.isEnabled ).to.be.false;
+				expect( replaceInput.infoText ).to.match( /^Tip: Find some text/ );
+			} );
```
Comment (maxbarnas) [inline packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js:1097]: This test fails when the Chrome window, that executes the tests, is not focused. @oleq, AFAIK, you created a workaround some time ago for a similar problem, right?

**Comment by maxbarnas** [comment]
Of all the handover tasks only the style linter issue seems to be fixed ( #9928 ).

**Comment by maxbarnas** [comment]
Apart from the remarks I posted above everything else looks great! 🎉  Once we know what to do with them, I am ready to merge.

File: packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js (hunk 2 above)
Comment (oleq) [reply packages/ckeditor5-find-and-replace/tests/ui/findandreplaceformview.js]: Resolved in 120d911.

**Comment by oleq** [comment]
> ### Steps \[SAFARI only\]:
> 
> 1.  Open F&R dropdown.
> 2.  Click on it.
> 
> ### Result:
> 
> There is a blue outline with an irregular shape:

I addressed this in df16832.

**Pull request merged**